lib/Models/Cesium.js

```diff
@@ -597,6 +599,161 @@ Cesium.prototype.lowerToBottom = function(item) {
+        }
+    }
+
+    return vectorFeatures;
+};
+
+/**
+ * Hooks into the {@link ImageryProvider#pickFeatures} method of every imagery provider in the scene - when this method is
+ * evaluated (usually as part of feature picking), it will record the tile coordinates used against the url of the
+ * imagery provider in an object that is returned by this method. Hooks are removed immediately after being executed once.
+ *
+ * @returns {{x, y, level}} A map of urls to the coords used by the imagery provider when picking features. Will
+ *     initially be empty but will be updated as the hooks are evaluated.
+ * @private
+ */
+Cesium.prototype._attachProviderCoordHooks = function() {
+    var providerCoords = {};
+
+    var pickFeaturesHook = function(imageryProvider, oldPick, x, y, level, longitude, latitude) {
+        var featuresPromise = oldPick.call(imageryProvider, x, y, level, longitude, latitude);
+
+        // Use url to uniquely identify providers because what else can we do?
+        if (imageryProvider.url) {
+            providerCoords[imageryProvider.url] = {
+                x: x,
+                y: y,
+                level: level
+            };
+        }
+
+        imageryProvider.pickFeatures = oldPick;
+        return featuresPromise;
+    };
+
+    for (var j = 0; j < this.scene.imageryLayers.length; j++) {
+        var imageryProvider = this.scene.imageryLayers.get(j).imageryProvider;
+        imageryProvider.pickFeatures = pickFeaturesHook.bind(undefined, imageryProvider, imageryProvider.pickFeatures);
+    }
+
+    return providerCoords;
+};
+
+/**
+ * Builds a {@link PickedFeatures} object from a number of inputs.
+ *
+ * @param {{x, y, level}} providerCoords A map of imagery provider urls to the coords used to get features for that provider.
+ * @param {Cartesian3} pickPosition The position in the 3D model that has been picked.
+ * @param {Entity[]} existingFeatures Existing features - the results of feature promises will be appended to this.
+ * @param {Promise[]} featurePromises Zero or more promises that will be combined as part of {@link PickedFeatures#allFeaturesAvailablePromise}
+ *     and their results used to build the final {@link PickedFeatures#features} array.
+ * @param {ImageryLayer[]} imageryLayers An array of ImageryLayers that should line up with the one passed as featurePromises.
+ * @param {number} defaultHeight The height to use for feature position heights if none is available when picking.
+ * @returns {PickedFeatures} A {@link PickedFeatures} object that is a combination of everything passed.
+ * @private
+ */
+Cesium.prototype._buildPickedFeatures = function(providerCoords, pickPosition, existingFeatures, featurePromises, imageryLayers, defaultHeight) {
+    var result = new PickedFeatures();
+
+    result.providerCoords = providerCoords;
+    result.pickPosition = pickPosition;
+
+    result.allFeaturesAvailablePromise = when.all(featurePromises).then(function(allFeatures) {
```

Comment:
Following on from my question about the changes to what `pickFeatures` returns... it's not obvious to me what sort of objects these `featurePromises` resolve to? Especially as they can come from `imageryProvider.pickFeatures` or `this.scene.imageryLayers.pickImageryLayerFeatures`...